openshift-kube-apiserver/admission/autoscaling/managementcpusoverride/admission.go

```diff
@@ -0,0 +1,294 @@
+	"k8s.io/klog/v2"
+	coreapi "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/kubelet/cm"
+)
+
+const (
+	PluginName                              = "autoscaling.openshift.io/ManagementCPUsOverride"
+	podWorkloadAnnotation                   = "workload.openshift.io"
+	podWorkloadAnnotationEffect             = "effect"
+	workloadEffectPreferredDuringScheduling = "PreferredDuringScheduling"
+	managedCapacityLabel                    = "management.workload.openshift.io/cores"
+	containerCPUSharesAnnotation            = "io.openshift.workload.management.cpushares"
+	containerCPUSetAnnotation               = "io.openshift.workload.management.cpuset"
+	namespaceAllowedAnnotation              = "workload.openshift.io/allowed"
+	workloadTypeManagement                  = "management"
+)
+
+var _ = initializer.WantsExternalKubeInformerFactory(&managementCPUsOverride{})
+var _ = admission.MutationInterface(&managementCPUsOverride{})
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &managementCPUsOverride{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+type managementCPUsOverride struct {
+	*admission.Handler
+	nsLister   corev1listers.NamespaceLister
+	nodeLister corev1listers.NodeLister
+}
+
+func (a *managementCPUsOverride) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	a.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	a.nodeLister = kubeInformers.Core().V1().Nodes().Lister()
+}
+
+func (a *managementCPUsOverride) ValidateInitialization() error {
+	if a.nsLister == nil {
+		return fmt.Errorf("%s did not get a namespace lister", PluginName)
+	}
+	if a.nodeLister == nil {
+		return fmt.Errorf("%s did not get a node lister", PluginName)
+	}
+	return nil
+}
+
+func (a *managementCPUsOverride) Admit(ctx context.Context, attr admission.Attributes, o admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil
+	}
+
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+
+	// allow annotations on project to override management pods CPUs requests
+	ns, err := a.nsLister.Get(attr.GetNamespace())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving namespace %q: %v", PluginName, attr.GetNamespace(), err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	nodes, err := a.nodeLister.List(labels.Everything())
+	if err != nil {
+		klog.Warningf("%s got an error retrieving nodes: %v", PluginName, err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	}
+
+	if !isManagementResourceAvailable(nodes) {
+		klog.Warningf("%s no nodes with resource %s is available", PluginName, managedCapacityLabel)
+		return nil
+	}
+
```

Comment:
In these cases where the pod is not allowed to use the partitioning feature, we need to strip any annotations that might trigger the underlying components to partition the workload. For example, that prevents a savvy user from adding the annotation with settings for cpushares that would normally be added by this admission hook. See https://github.com/openshift/enhancements/pull/703/files#diff-bdf426b496fe2c7122db9eeaf303b30ff4561b4f7ac157c21996edff1e67e239R517-R520

Reply:
done